File: backend/src/application/use-cases/LoginUser.ts
import { IUserRepository } from '../../domain/repositories/IUserRepository';
import { JwtService } from '../../infrastructure/auth/jwt';

export interface LoginUserDTO {
  email: string;
  password: string;
}

export interface IPasswordVerifier {
  compare(password: string, hash: string): Promise<boolean>;
}

export interface LoginResult {
  token: string;
  user: {
    id: string;
    email: string;
    fullName: string;
    role: string;
    communityId?: string;
  };
}

export class LoginUser {
  constructor(
    private readonly userRepository: IUserRepository,
    private readonly passwordVerifier: IPasswordVerifier,
    private readonly jwtService: JwtService
  ) {}

  async execute(dto: LoginUserDTO): Promise<LoginResult> {
    const user = await this.userRepository.findByEmail(dto.email);
    if (!user) {
      throw new Error('Invalid email or password');
    }

    const valid = await this.passwordVerifier.compare(dto.password, user.passwordHash);
    if (!valid) {
      throw new Error('Invalid email or password');
    }

    const token = this.jwtService.sign({
      userId: user.id,
      email: user.email,
      role: user.role,
    });

    return {
      token,
      user: {
        id: user.id,
        email: user.email,
        fullName: user.fullName,
        role: user.role,
        communityId: user.communityId,
      },
    };
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/LoginUser.cs
using Hestia.Application.DTOs;
using Hestia.Application.Interfaces;
using Hestia.Domain.Repositories;

namespace Hestia.Application.UseCases;

public class LoginUser
{
    private readonly IUserRepository _userRepository;
    private readonly IPasswordHasher _passwordHasher;
    private readonly IJwtService _jwtService;

    public LoginUser(IUserRepository userRepository, IPasswordHasher passwordHasher, IJwtService jwtService)
    {
        _userRepository = userRepository;
        _passwordHasher = passwordHasher;
        _jwtService = jwtService;
    }

    public async Task<LoginResponse> ExecuteAsync(LoginRequest dto, CancellationToken ct = default)
    {
        var user = await _userRepository.FindByEmailAsync(dto.Email, ct);
        if (user is null || !_passwordHasher.Verify(dto.Password, user.PasswordHash))
            throw new ApplicationException("Invalid email or password");

        var token = _jwtService.GenerateToken(user.Id, user.Email, user.Role.ToString().ToUpperInvariant());

        return new LoginResponse
        {
            Token = token,
            User = new UserSummary
            {
                Id = user.Id,
                Email = user.Email,
                FullName = user.FullName,
                Role = user.Role.ToString().ToUpperInvariant(),
                CommunityId = user.CommunityId
            }
        };
    }
}
